Geometry/RPCGeometryBuilder/src/RPCGeometryBuilderFromCondDB.cc

```diff
@@ -89,76 +78,113 @@ RPCGeometry* RPCGeometryBuilderFromCondDB::build(const RecoIdealGeometry& rgeo)
       //      if (tran[2] > 0. )
       newY *= -1;
       Basic3DVector<float> newZ(0.,1.,0.);
-      rot.rotateAxes (newX, newY,newZ);	
+      rot.rotateAxes (newX, newY,newZ);
     }
-    
+
     BoundPlane* bp = new BoundPlane(pos,rot,bounds);
     ReferenceCountingPointer<BoundPlane> surf(bp);
-    RPCRoll* r=new RPCRoll(rpcid,surf,rollspecs);
+    RPCRoll* r = new RPCRoll(rpcid,surf,rollspecs);
     geometry->add(r);
-    
-    std::list<RPCRoll *> rls;
-    if (chids.find(chid)!=chids.end()){
-      rls = chids[chid];
-    }
-    rls.push_back(r);
-    chids[chid]=rls;
+
+    auto rls = chids.find(chid);
+    if ( rls == chids.end() ) rls = chids.insert(std::make_pair(chid, std::set<RPCRoll*>())).first;
+    rls->second.insert(r);
   }
-  // Create the RPCChambers and store them on the Geometry 
-  for( std::map<RPCDetId, std::list<RPCRoll *> >::iterator ich=chids.begin();
-       ich != chids.end(); ich++){
-    RPCDetId chid = ich->first;
-    std::list<RPCRoll * > rls = ich->second;
-
-    // compute the overall boundplane. At the moment we use just the last
-    // surface
+
+  // Create the RPCChambers and store them on the Geometry
+  for ( auto ich=chids.begin(); ich != chids.end(); ++ich ) {
+    const RPCDetId& chid = ich->first;
+    const auto& rls = ich->second;
+
+    // compute the overall boundplane.
     BoundPlane* bp=0;
-    for(std::list<RPCRoll *>::iterator rl=rls.begin();
-    rl!=rls.end(); rl++){
-      const BoundPlane& bps = (*rl)->surface();
-      bp = const_cast<BoundPlane *>(&bps);
+    if ( !rls.empty() ) {
+      // First set the baseline plane to calculate relative poisions
+      const auto& refSurf = (*rls.begin())->surface();
+      if ( chid.region() == 0 ) {
+        float corners[6] = {0,0,0,0,0,0};
+        for ( auto rl : rls ) {
+          const double h2 = rl->surface().bounds().length()/2;
+          const double w2 = rl->surface().bounds().width()/2;
+          const auto x1y1AtRef = refSurf.toLocal(rl->toGlobal(LocalPoint(-w2,-h2,0)));
+          const auto x2y2AtRef = refSurf.toLocal(rl->toGlobal(LocalPoint(+w2,+h2,0)));
+          corners[0] = std::min(corners[0], x1y1AtRef.x());
+          corners[1] = std::min(corners[1], x1y1AtRef.y());
+          corners[2] = std::max(corners[2], x2y2AtRef.x());
+          corners[3] = std::max(corners[3], x2y2AtRef.y());
+
+          corners[4] = std::min(corners[4], x1y1AtRef.z());
+          corners[5] = std::max(corners[5], x1y1AtRef.z());
+        }
+        const LocalPoint lpOfCentre((corners[0]+corners[2])/2, (corners[1]+corners[3])/2, 0);
+        const auto gpOfCentre = refSurf.toGlobal(lpOfCentre);
+        auto bounds = new RectangularPlaneBounds((corners[2]-corners[0])/2, (corners[3]-corners[1])/2, (corners[5]-corners[4])+0.5);
+        bp = new BoundPlane(gpOfCentre, refSurf.rotation(), bounds);
+      }
+      else {
+        float cornersLo[3] = {0,}, cornersHi[3] = {0,};
+        float cornersZ[2] = {0,};
```

Comment:
are the higher (above 0) indices intentionally uninitialized?
You will likely get values from the previous calls in cornersLo[1, 2] etc